featuregates/controller/pkg/feature/feature_suite_test.go

```diff
@@ -0,0 +1,565 @@
+	return nil
+}
+
+var _ = BeforeSuite(func() {
+	logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
+
+	ctx, cancel = context.WithCancel(context.TODO())
+
+	By("bootstrapping test environment")
+	testEnv = &envtest.Environment{
+		CRDDirectoryPaths:     []string{filepath.Join("..", "..", "..", "..", "apis", "core", "config", "crd", "bases")},
+		ErrorIfCRDPathMissing: true,
+	}
+
+	var err error
+	cfg, err = testEnv.Start()
+	Expect(err).NotTo(HaveOccurred())
+	Expect(cfg).NotTo(BeNil())
+
+	testEnv.ControlPlane.APIServer.Configure().Append("admission-control", "MutatingAdmissionWebhook")
+	testEnv.ControlPlane.APIServer.Configure().Append("admission-control", "ValidatingAdmissionWebhook")
+
+	err = corev1alpha2.AddToScheme(scheme.Scheme)
+	Expect(err).NotTo(HaveOccurred())
+
+	k8sManager, err := ctrl.NewManager(cfg, ctrl.Options{
+		Scheme:             scheme.Scheme,
+		MetricsBindAddress: "0",
+		Host:               "127.0.0.1",
+		Port:               9443,
+	})
+	Expect(err).ToNot(HaveOccurred())
+
+	err = generateCertificateAndManifests()
+	Expect(err).ToNot(HaveOccurred())
+
+	k8sManager.GetWebhookServer().TLSMinVersion = "1.2"
+	k8sManager.GetWebhookServer().CertDir = tmpDir
+	k8sManager.GetWebhookServer().CertName = "feature-webhook.crt"
+	k8sManager.GetWebhookServer().KeyName = "feature-webhook.key"
+	k8sManager.GetWebhookServer().Port = 9443
+
+	k8sClient, err = client.New(cfg, client.Options{Scheme: scheme.Scheme})
+	Expect(err).NotTo(HaveOccurred())
+	Expect(k8sClient).NotTo(BeNil())
+
+	dynamicClient, err := dynamic.NewForConfig(cfg)
+	Expect(err).ToNot(HaveOccurred())
+	Expect(dynamicClient).ToNot(BeNil())
+
+	err = (&FeatureReconciler{
+		Client: k8sManager.GetClient(),
+		Scheme: k8sManager.GetScheme(),
+		Log:    setupLog,
+	}).SetupWithManager(k8sManager)
+	Expect(err).ToNot(HaveOccurred())
+
+	ns := &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "tkg-system"}}
+	Expect(k8sClient.Create(ctx, ns)).To(Succeed())
+
+	err = (&corev1alpha2.FeatureGate{}).SetupWebhookWithManager(k8sManager)
+	Expect(err).ToNot(HaveOccurred())
+
+	go func() {
+		defer GinkgoRecover()
+		err = k8sManager.Start(ctx)
+		Expect(err).ToNot(HaveOccurred(), "failed to run manager")
+	}()
+
+	err = createResources(cfg, dynamicClient)
+	Expect(err).ToNot(HaveOccurred())
+})
+
+var _ = AfterSuite(func() {
+	cancel()
+	By("tearing down the test environment")
+	err := testEnv.Stop()
+	Expect(err).NotTo(HaveOccurred())
+})
+
+func getTestFeatureGate() *corev1alpha2.FeatureGate {
+	randomNumber, err := rand.Int(rand.Reader, big.NewInt(100000))
+	if err != nil {
+		return nil
+	}
+
+	return &corev1alpha2.FeatureGate{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fmt.Sprintf("featuregate-%v", randomNumber),
+		},
+		Spec: corev1alpha2.FeatureGateSpec{
+			Features: []corev1alpha2.FeatureReference{},
+		},
+		Status: corev1alpha2.FeatureGateStatus{},
+	}
+}
+
+func getTestFeature() *corev1alpha2.Feature {
```

Comment:
Why is this required?

Reply:
so that you don't have to do [explicitly set the stability level](https://github.com/vmware-tanzu/tanzu-framework/pull/4453/files#diff-d067ce787aa01aa3bc0f7055d55a60592af39dd232a5e3697e1b8bfb1222a8f7R303) in each test case, you can just pass the stability level to the function and the function will return feature with that stability level, it also makes sense to rename this function to `getTestFeatureWithStabilityLevel` or something similar

Reply:
Done